component/common/loki/client/internal/marker_file_handler.go

```diff
@@ -0,0 +1,87 @@
+package internal
+
+import (
+	"fmt"
+	"github.com/grafana/agent/component/common/loki/wal"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/prometheus/tsdb/fileutil"
+)
+
+type MarkerFileHandler interface {
+	wal.Marker
+	MarkSegment(segment int)
+}
+
+type markerFileHandler struct {
+	logger log.Logger
+
+	lastMarkedSegmentFilePath string
+}
+
+var (
+	_ MarkerFileHandler = (*markerFileHandler)(nil)
+)
+
+func NewMarkerFileHandler(logger log.Logger, walDir string) (MarkerFileHandler, error) {
+	markerDir := filepath.Join(walDir, "remote")
+	// attempt to create dir if doesn't exist
+	if err := os.MkdirAll(markerDir, 0o777); err != nil {
+		return nil, fmt.Errorf("error creating segment marker folder %q: %w", markerDir, err)
+	}
+
+	mfh := &markerFileHandler{
+		logger:                    logger,
+		lastMarkedSegmentFilePath: filepath.Join(markerDir, "segment"),
+	}
+
+	return mfh, nil
+}
+
+// LastMarkedSegment implements wlog.Marker.
+func (mfh *markerFileHandler) LastMarkedSegment() int {
```

Comment:
Yes, I agree when we introduce offsets into the segment (so we can keep more granular markers, and seek directly to that byte) we might want to change the file structure. To keep this PR as small as possible, I'm down for leaving this as is, and later we can introduce some file headers in case we need to change the marker files after offsets as well.

Reply:
Ok - I'm happy to leave it as it is for now. I agree that this is a reasonable first implementation.